spotify_tensorflow/luigi/python_dataflow_task.py

```diff
@@ -0,0 +1,303 @@
+        staging_location = "gs://gcp-project-playground/user/staging"
+        temp_location = "gs://gcp-project-playground/user/tmp"
+        max_num_workers = 20
+        region = "europe-west1"
+        service_account_email = "service_account@gcp-project.iam.gserviceaccount.com"
+
+        def output(self):
+            ...
+    """
+    # Required dataflow args
+    python_script = None
+    project = None
+    staging_location = None
+
+    # Dataflow requires one and only one of:
+    zone = None
+    region = None
+
+    # Optional dataflow args
+    temp_location = None
+    num_workers = None
+    autoscaling_algorithm = None
+    max_num_workers = None
+    network = None
+    subnetwork = None
+    disk_size_gb = None
+    worker_machine_type = None
+    worker_disk_type = None
+    service_account = None
+    job_name = None
+
+    # requirements_file for workers
+    requirements_file = None
+
+    local_runner = False
+
+    def __init__(self, *args, **kwargs):
+        super(PythonDataflowTask, self).__init__(*args, **kwargs)
+        self._output = self.output()
+        if isinstance(self._output, luigi.Target):
+            self._output = {"output": self._output}
+
+    def on_successful_run(self):
+        """ Callback that gets called right after the dataflow job has finished successfully but
+        before validate_output is run.
+        """
+        pass
+
+    def validate_output(self):
+        """ Callback that can be used to validate your output before it is moved to it's final
+        location. Returning false here will cause the job to fail, and output to be removed instead
+        of published.
+
+        :return: Whether the output is valid or not
+        :rtype: Boolean
+        """
+        return True
+
+    def file_pattern(self):
+        """ If one/some of the input target files are not in the pattern of part-*,
+        we can add the key of the required target and the correct file pattern
+        that should be appended in the command line here. If the input target key is not found
+        in this dict, the file pattern will be assumed to be part-* for that target.
+
+        :return A dictionary of overrided file pattern that is not part-* for the inputs
+        :rtype: Dict of String to String
+        """
+        return {}
+
+    def run(self):
+        cmd_line = self._mk_cmd_line()
+        logger.info(" ".join(cmd_line))
+
+        try:
+            run_with_logging(cmd_line)
+        except subprocess.CalledProcessError as e:
+            logging.error(e, exc_info=True)
+            # exit luigi with the same exit code as the python dataflow job proccess
+            # In this way users can easily exit the job with code 50 to avoid Styx retries
+            # https://github.com/spotify/styx/blob/master/doc/design-overview.md#workflow-state-graph
+            os._exit(e.returncode)
+
+        self.on_successful_run()
+        if self.validate_output():
+            self._publish_outputs()
```

Comment:
I included `_publish_outputs` since I'm now doing a dynamic check by `hasattr(target, "publish")` to work around. Also want to have a placeholder for `self.validate_output()` check. It may be useful when we want to add tfdv validation later.